### PR TITLE
Improve UT coverage for IBMPowerVSCluster controller

### PR DESCRIPTION
**Pull request by Karthik-K-N**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # Related to https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/1818

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

File: controllers/ibmpowervscluster_controller_test.go [hunk 1]
```diff
@@ -433,6 +441,341 @@ func TestIBMPowerVSClusterReconciler_delete(t *testing.T) {
...
+				cluster: clusterScope.IBMPowerVSCluster,
+			}
+			wg := &sync.WaitGroup{}
+			wg.Add(1)
+			reconciler.reconcileVPCResources(clusterScope, pvsCluster, ch, wg)
+			wg.Wait()
+			close(ch)
+			g.Expect(<-ch).To(Equal(tc.reconcileResult))
+			if len(tc.conditions) > 0 {
+				ignoreLastTransitionTime := cmp.Transformer("", func(metav1.Time) metav1.Time {
```
Comment (Amulyam24) [inline controllers/ibmpowervscluster_controller_test.go:742]: For my understanding, why are we adding this condition?

File: controllers/ibmpowervscluster_controller_test.go [hunk 2]
```diff
@@ -433,6 +441,341 @@ func TestIBMPowerVSClusterReconciler_delete(t *testing.T) {
...
+					Status:             "False",
+					Severity:           capiv1beta1.ConditionSeverityError,
+					LastTransitionTime: metav1.Time{},
+					Reason:             infrav1beta2.VPCReconciliationFailedReason,
+					Message:            "vpc not found",
+				},
+			},
+		},
+		{
+			name: "when ReconcileVPC returns requeue",
```
Comment (Amulyam24) [inline controllers/ibmpowervscluster_controller_test.go:477]:
```suggestion
			name: "when ReconcileVPC returns requeue as true",
```
Almost every reconcile returns requeue, better to change statement wherever applicable.

File: controllers/ibmpowervscluster_controller_test.go [hunk 3]
```diff
@@ -433,6 +441,341 @@ func TestIBMPowerVSClusterReconciler_delete(t *testing.T) {
...
+					Reason:             infrav1beta2.LoadBalancerReconciliationFailedReason,
+					Message:            "loadbalancer not found",
+				},
+				getVPCReadyCondition(),
+				getVPCSGReadyCondition(),
+				getVPCSubnetReadyCondition(),
+			},
+		},
+		{
+			name: "when Reconciling LoadBalancer and loadbalancer is ready",
```
Comment (Amulyam24) [inline controllers/ibmpowervscluster_controller_test.go:679]:
```suggestion
			name: "when Reconciling LoadBalancer returns with loadbalancer status as ready",
```

File: controllers/ibmpowervscluster_controller_test.go (hunk 2 above)
Comment (Karthik-K-N) [reply controllers/ibmpowervscluster_controller_test.go]: Updated as suggested

File: controllers/ibmpowervscluster_controller_test.go (hunk 3 above)
Comment (Karthik-K-N) [reply controllers/ibmpowervscluster_controller_test.go]: Updated as suggested

File: controllers/ibmpowervscluster_controller_test.go (hunk 1 above)
Comment (Karthik-K-N) [reply controllers/ibmpowervscluster_controller_test.go:1568]: If you mean why cmp.Transfer() it simply ignores the LastTransition time as we cannot compare two time.Now() values.

File: controllers/ibmpowervscluster_controller_test.go [hunk 4]
```diff
@@ -294,24 +309,318 @@ func TestIBMPowerVSClusterReconciler_Reconcile(t *testing.T) {
...
+				}
+				mockPowerVS := powervsmock.NewMockPowerVS(gomock.NewController(t))
+				mockPowerVS.EXPECT().GetDatacenterCapabilities(gomock.Any()).Return(map[string]bool{"power-edge-router": true}, nil)
+				clusterScope.IBMPowerVSClient = mockPowerVS
+				return clusterScope
+			},
+			expectedError: errors.New("resource group name is not set"),
+		},
+		{
+			name: "When reconcile Power VS resource returns requeue",
```
Comment (Amulyam24) [inline controllers/ibmpowervscluster_controller_test.go:374]:
```suggestion
			name: "When reconcile Power VS resource returns requeue as true",
```
Please change wherever applicable

File: controllers/ibmpowervscluster_controller_test.go [hunk 5]
```diff
@@ -294,24 +309,318 @@ func TestIBMPowerVSClusterReconciler_Reconcile(t *testing.T) {
...
+				mockVPC := vpcmock.NewMockVpc(gomock.NewController(t))
+				mockVPC.EXPECT().GetVPCByName(gomock.Any()).Return(nil, errors.New("vpc not found"))
+				clusterScope.IBMVPCClient = mockVPC
+
+				return clusterScope
+			},
+			expectedResult: ctrl.Result{RequeueAfter: 30 * time.Second},
+		},
+		{
+			name: "When reconcile VPC and Power VS resource returns error",
```
Comment (Amulyam24) [inline controllers/ibmpowervscluster_controller_test.go:416]: maybe we can also cover the case where only requeue with true is returned without an error?

File: controllers/ibmpowervscluster_controller_test.go [hunk 6]
```diff
@@ -294,24 +309,318 @@ func TestIBMPowerVSClusterReconciler_Reconcile(t *testing.T) {
...
+				mockVPC := vpcmock.NewMockVpc(gomock.NewController(t))
+				mockVPC.EXPECT().GetVPCByName(gomock.Any()).Return(nil, errors.New("vpc not found"))
+				clusterScope.IBMVPCClient = mockVPC
+
+				return clusterScope
+			},
+			expectedError: kerrors.NewAggregate([]error{errors.New("error getting resource instance"), errors.New("vpc not found")}),
+		},
+		{
+			name: "When reconcile TransitGateway returns error",
```
Comment (Amulyam24) [inline controllers/ibmpowervscluster_controller_test.go:454]: any reason for not verifying the conditions for TG and COS?

**Comment by Amulyam24** [comment]
/retitle Improve UT coverage for IBMPowerVSCluster controller

File: controllers/ibmpowervscluster_controller_test.go (hunk 1 above)
Comment (Amulyam24) [reply controllers/ibmpowervscluster_controller_test.go:1050]: I'm trying to understand why `g.Expect(pvsCluster.cluster.GetConditions()).To(Equal(tc.conditions))` not work?

Comment (Karthik-K-N) [reply controllers/ibmpowervscluster_controller_test.go:1050]: Yes, It works, I will update.

File: controllers/ibmpowervscluster_controller_test.go (hunk 5 above)
Comment (Karthik-K-N) [reply controllers/ibmpowervscluster_controller_test.go:462]: Added a new test case where both vpc and power vs resources returns reque as true without error

File: controllers/ibmpowervscluster_controller_test.go (hunk 6 above)
Comment (Karthik-K-N) [reply controllers/ibmpowervscluster_controller_test.go:500]: Verified both TG and COS conidtions.

**Review by Amulyam24 — APPROVED** [review]
/lgtm

File: controllers/ibmpowervscluster_controller_test.go [hunk 7]
```diff
@@ -294,24 +309,397 @@ func TestIBMPowerVSClusterReconciler_Reconcile(t *testing.T) {
...
+						ObjectMeta: metav1.ObjectMeta{
+							Finalizers: []string{infrav1beta2.IBMPowerVSClusterFinalizer},
+						},
+					},
+				}
+			},
+			clusterStatus: true,
+		},
+		{
+			name: "When Power VS zone does not support PER",
```
Comment (Prajyot-Parab) [inline controllers/ibmpowervscluster_controller_test.go:340]:
```suggestion
			name: "When PowerVS zone does not support PER",
```

File: controllers/ibmpowervscluster_controller_test.go [hunk 8]
```diff
@@ -294,24 +309,397 @@ func TestIBMPowerVSClusterReconciler_Reconcile(t *testing.T) {
...
+				}
+				mockPowerVS := powervsmock.NewMockPowerVS(gomock.NewController(t))
+				mockPowerVS.EXPECT().GetDatacenterCapabilities(gomock.Any()).Return(map[string]bool{"power-edge-router": true}, nil)
+				clusterScope.IBMPowerVSClient = mockPowerVS
+				return clusterScope
+			},
+			expectedError: errors.New("resource group name is not set"),
+		},
+		{
+			name: "When reconcile Power VS resource returns requeue as true",
```
Comment (Prajyot-Parab) [inline controllers/ibmpowervscluster_controller_test.go:375]:
```suggestion
			name: "When reconcile PowerVS resource returns requeue as true",
```

File: controllers/ibmpowervscluster_controller_test.go [hunk 9]
```diff
@@ -294,24 +309,397 @@ func TestIBMPowerVSClusterReconciler_Reconcile(t *testing.T) {
...
+				mockVPC := vpcmock.NewMockVpc(gomock.NewController(t))
+				mockVPC.EXPECT().GetVPCByName(gomock.Any()).Return(nil, errors.New("vpc not found"))
+				clusterScope.IBMVPCClient = mockVPC
+
+				return clusterScope
+			},
+			expectedResult: ctrl.Result{RequeueAfter: 30 * time.Second},
+		},
+		{
+			name: "When reconcile Power VS and VPC resource returns requeue as true",
```
Comment (Prajyot-Parab) [inline controllers/ibmpowervscluster_controller_test.go:417]:
```suggestion
			name: "When reconcile PowerVS and VPC resource returns requeue as true",
```

File: controllers/ibmpowervscluster_controller_test.go [hunk 10]
```diff
@@ -294,24 +309,397 @@ func TestIBMPowerVSClusterReconciler_Reconcile(t *testing.T) {
...
+				mockVPC := vpcmock.NewMockVpc(gomock.NewController(t))
+				mockVPC.EXPECT().GetVPC(gomock.Any()).Return(&vpcv1.VPC{Status: ptr.To("pending")}, nil, nil)
+				clusterScope.IBMVPCClient = mockVPC
+
+				return clusterScope
+			},
+			expectedResult: ctrl.Result{RequeueAfter: 30 * time.Second},
+		},
+		{
+			name: "When reconcile VPC and Power VS resource returns error",
```
Comment (Prajyot-Parab) [inline controllers/ibmpowervscluster_controller_test.go:462]:
```suggestion
			name: "When reconcile VPC and PowerVS resource returns error",
```

File: controllers/ibmpowervscluster_controller_test.go [hunk 11]
```diff
@@ -294,24 +309,397 @@ func TestIBMPowerVSClusterReconciler_Reconcile(t *testing.T) {
...
+				return &scope.PowerVSClusterScope{
+					IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+						ObjectMeta: metav1.ObjectMeta{
+							Finalizers:  []string{infrav1beta2.IBMPowerVSClusterFinalizer},
+							Annotations: map[string]string{infrav1beta2.CreateInfrastructureAnnotation: "true"},
+						},
 					},
+				}
+			},
+			expectedError: errors.New("PowerVS zone is not set"),
```
Comment (Prajyot-Parab) [inline controllers/ibmpowervscluster_controller_test.go:351]: wondering if the expected error is inline with expectation of the test case (as per name)
`name: "When Power VS zone does not support PER",`

**Review by Prajyot-Parab — CHANGES_REQUESTED** [review]

File: controllers/ibmpowervscluster_controller_test.go (hunk 11 above)
Comment (Karthik-K-N) [reply controllers/ibmpowervscluster_controller_test.go]: Updated the test case.

File: controllers/ibmpowervscluster_controller_test.go (hunk 7 above)
Comment (Karthik-K-N) [reply controllers/ibmpowervscluster_controller_test.go]: Updated at all the relevant places as suggested.

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm

**Pull request merged**